Userland/Libraries/LibELF/DynamicLinker.cpp

```diff
@@ -594,6 +600,33 @@ static Result<void, DlErrorMessage> __dladdr(void const* addr, Dl_info* info)
     return {};
 }
 
+static void __call_fini_functions()
+{
+    typedef void (*FiniFunc)();
+    auto const* object = *s_global_objects.get(s_main_program_path);
+
+    if (object->has_fini_array_section()) {
+        auto fini_array_section = object->fini_array_section();
+
+        FiniFunc* fini_begin = (FiniFunc*)(fini_array_section.address().as_ptr());
+        FiniFunc* fini_end = fini_begin + fini_array_section.entry_count();
+        while (fini_begin != fini_end) {
+            --fini_end;
+
+            // Android sources claim that these can be -1, to be ignored.
+            // 0 deffiniely shows up. Apparently 0/-1 are valid? Confusing.
+            if (!*fini_end || ((FlatPtr)*fini_end == (FlatPtr)-1))
+                continue;
+            (*fini_end)();
+        }
+    }
+
+    if (object->has_fini_section()) {
+        auto fini_function = object->fini_section_function();
+        (fini_function)();
+    }
+}
+
```

Comment:
Shouldn't we also be calling the finalizers defined in other loaded objects as well, not just the main program (c.f. `DynamicLoader::call_object_init_functions`)?

Reply:
Would iterating in reverse topological order work? `OrderedHashMap` uses a doubly linked list internally, so it should be possible, we just need to implement a reverse iterator for it.

Reply:
I could have a go tonight if you'd prefer.

Reply:
#21219